src/rgw/rgw_trim_bilog.cc

```diff
@@ -447,29 +451,70 @@ int BucketTrimInstanceCR::operate()
   reenter(this) {
     ldout(cct, 4) << "starting trim on bucket=" << bucket_instance << dendl;
 
+    get_policy_params.zone = zone_id;
+    get_policy_params.bucket = bucket;
+    yield call(new RGWBucketGetSyncPolicyHandlerCR(store->svc()->rados->get_async_processor(),
+                                                   store,
+                                                   get_policy_params,
+                                                   source_policy));
+    if (retcode < 0) {
+      if (retcode != -ENOENT) {
+        ldout(cct, 0) << "ERROR: failed to fetch policy handler for bucket=" << bucket << dendl;
+      }
+
+      return set_cr_error(retcode);
+    }
+
+    if (auto& opt_bucket_info = source_policy->policy_handler->get_bucket_info();
+        opt_bucket_info) {
+      pbucket_info = &(*opt_bucket_info);
+    } else {
+      /* this shouldn't really happen */
+      return set_cr_error(-ENOENT);
+    }
+
     // query peers for sync status
-    set_status("fetching sync status from peers");
+    set_status("fetching sync status from relevant peers");
     yield {
-      // query data sync status from each sync peer
-      rgw_http_param_pair params[] = {
-        { "type", "bucket-index" },
-        { "status", nullptr },
-        { "bucket", bucket_instance.c_str() },
-        { "source-zone", zone_id.c_str() },
-        { nullptr, nullptr }
-      };
+      const auto& all_dests = source_policy->policy_handler->get_all_dests();
+
+      vector<rgw_zone_id> zids;
+      rgw_zone_id last_zid;
+      for (auto& diter : all_dests) {
+        const auto& zid = diter.first;
+        if (zid == last_zid) {
+          continue;
+        }
+        last_zid = zid;
+        zids.push_back(zid);
+      }
+
+      peer_status.resize(zids.size());
+
+      auto& zone_conn_map = store->svc()->zone->get_zone_conn_map();
 
       auto p = peer_status.begin();
-      for (auto& c : store->svc()->zone->get_zone_data_notify_to_map()) {
+      for (auto& zid : zids) {
+        // query data sync status from each sync peer
+        rgw_http_param_pair params[] = {
+          { "type", "bucket-index" },
+          { "status", nullptr },
+          { "options", "merge" },
+          { "source-bucket", bucket_instance.c_str() },
```

Comment:
a nautilus cluster won't recognize this `source-bucket` param, so i don't think we can rename it from `bucket`

(it also won't recognize the `options=merge` param, but that's fine because it won't have multiple targets and trim will work as is)

Reply:
It isn't really a rename, it's a different param. Can put both params and ignore it if `options==merge`

Reply:
on second thought, can drop `source-bucket` here, but keep the api accepting it as a param. Is treated as `source-bucket` in case it isn't provided when `options==merge`.

Reply:
on second thought, can drop source-bucket here, but keep the api accepting it as a param. Is treated as source-bucket in case it isn't provided when options==merge.

Reply:
https://github.com/ceph/ceph/pull/31686/commits/fc2c939eeed199ab366d6e30e1a045e48234a556